File: src/main/cpp/2021/02/AoC2021_02.cpp
#include "assert.h"
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <set>
#include "../../aocd/aocd.hpp"

using namespace std;

const vector<string> TEST = {
    "forward 5",
    "down 5",
    "forward 8",
    "up 3",
    "down 8",
    "forward 2"
};

const string FORWARD = "forward";
const string UP = "up";
const string DOWN = "down";
const set<string> ALLOWED_COMMANDS = { FORWARD, UP, DOWN };

class Command {
    public:
        static Command* create(const string direction, const string amount) {
            return new Command(direction, stoi(amount));
        }

        string getDirection() const {
            return this->direction;
        }

        unsigned int getAmount() const {
            return this->amount;
        }

    private:
        string direction;
        unsigned int amount;

        Command(const string direction, const unsigned int amount) {
            if (ALLOWED_COMMANDS.find(direction) == ALLOWED_COMMANDS.end()) {
                throw "Illegal value";
            }
            this->direction = direction;
            this-> amount = amount;
        }
};

vector<Command> parse(const vector<string> &input) {
    vector<Command> commands;
    for (const string line: input) {
        vector<string> splits;
        istringstream iss(line);
        do {
            string split;
            iss >> split;
            splits.push_back(split);
        } while(iss);
        // assert splits.size() == 2
        commands.push_back(*Command::create(splits[0], splits[1]));
    }
    return commands;
}

int part1(const vector<string> &input) {
    vector<Command> commands = parse(input);
    int hor = 0, ver = 0;
    for (const Command command : commands) {
        if (command.getDirection() == UP) {
            ver -= command.getAmount();
        } else if (command.getDirection() == DOWN) {
            ver += command.getAmount();
        } else {
            hor += command.getAmount();
        }
    }
    return hor * ver;
}

int part2(const vector<string> &input) {
    vector<Command> commands = parse(input);
    int hor = 0, ver = 0, aim = 0;
    for (const Command command : commands) {
        if (command.getDirection() == UP) {
            aim -= command.getAmount();
        } else if (command.getDirection() == DOWN) {
            aim += command.getAmount();
        } else {
            hor += command.getAmount();
            ver += aim * command.getAmount();
        }
    }
    return hor * ver;
}


int main() {
    assert(part1(TEST) == 150);
    assert(part2(TEST) == 900);

    const vector<string> input = aocd::puzzle::getInputData(2021, 2);
    cout << "Part 1: " << part1(input) << endl;
    cout << "Part 2: " << part2(input) << endl;

    return 0;
}

File: src/main/java/AoC2021_02.java
import static java.util.stream.Collectors.toList;

import java.util.List;
import java.util.function.Function;

import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2021_02 extends AoCBase {
    
    // private static final String FORWARD = "forward";
    private static final String UP = "up";
    private static final String DOWN = "down";

    private final List<Command> commands;
	
	private AoC2021_02(final List<String> input, final boolean debug) {
		super(debug);
		final Function<String, Command> createCommand = s -> {
            final String[] split = s.split(" ");
            final String direction = split[0];
            return Command.create(direction, split[1]);
        };
        this.commands = input.stream().map(createCommand).collect(toList());
	}
	
	public static final AoC2021_02 create(final List<String> input) {
		return new AoC2021_02(input, false);
	}

	public static final AoC2021_02 createDebug(final List<String> input) {
		return new AoC2021_02(input, true);
	}
	
	@Override
    public Long solvePart1() {
	    long hor = 0, ver = 0;
	    for (final Command command : this.commands) {
	        if (UP.equals(command.direction())) {
	            ver -= command.amount();
	        } else if (DOWN.equals(command.direction())) {
	            ver += command.amount();
	        } else {
	            hor += command.amount();
	        }
        }
	    return hor * ver;
	}
	
	@Override
	public Long solvePart2() {
	    long hor = 0, ver = 0, aim = 0;
	    for (final Command command : this.commands) {
	        if (UP.equals(command.direction())) {
	            aim -= command.amount();
	        } else if (DOWN.equals(command.direction())) {
	            aim += command.amount();
	        } else {
	            hor += command.amount();
	            ver += (aim * command.amount());
	        }
        }
	    return hor * ver;
	}
	
	public static void main(final String[] args) throws Exception {
		assert AoC2021_02.createDebug(TEST).solvePart1() == 150;
		assert AoC2021_02.createDebug(TEST).solvePart2() == 900;
		
        final Puzzle puzzle = Aocd.puzzle(2021, 2);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2021_02.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2021_02.create(inputData)::solvePart2)
        );
	}
	
	private static final List<String> TEST = splitLines(
	        "forward 5\r\n" +
            "down 5\r\n" +
            "forward 8\r\n" +
            "up 3\r\n" +
            "down 8\r\n" +
            "forward 2"
    );
	
	record Command(String direction, int amount) {
	    
	    public static Command create(final String direction, final String amount) {
	        return new Command(direction, Integer.parseInt(amount));
	    }
	}
}
